### ISSUE TITLE
Thermostat Capability UI

### ISSUE THREAD

**Issue by benfrancis** [mozilla-iot/gateway#1629]

**Comment by benfrancis** [comment]
See also: https://github.com/mozilla-iot/schemas/issues/13

**Comment by benfrancis** [comment]
See also: https://github.com/mozilla-iot/schemas/issues/13

**Comment by benfrancis** [comment]
First pass on Thermostat UI design:
![thermostat_capability](https://user-images.githubusercontent.com/552417/65975571-b66e7480-e466-11e9-8af8-92892e6bbb42.png)

This fairly literal rendering of the properties is a bit clunky though, so I'm experimenting with a more custom UI design for the detail view, unlike anything we've done before (needs some work).
![thermostat_capability_custom](https://user-images.githubusercontent.com/552417/65975691-e4ec4f80-e466-11e9-8b9d-6b835087fcd2.png)

**Comment by mrstegeman** [comment]
One other thing we didn't discuss is the heating icon. Another option to the bacon icon is a flame. Just a thought. That's what the [Zen Thermostat](http://www.connectedcrib.com/wp-content/uploads/2014/09/zen.gif) does.

**Comment by benfrancis** [comment]
Here's a mockup using flame to mean heat, demonstrating two setpoints and a mini thermostat icon that looks slightly different to the temperature sensor one!

![thermostat_capability](https://user-images.githubusercontent.com/552417/66334340-cbe51200-e930-11e9-8cee-fc2f52c523a6.png)

I'm still toying with a more custom UI for thermostats which I quite like:
![thermostat_capability_custom](https://user-images.githubusercontent.com/552417/66334033-23cf4900-e930-11e9-9257-b34e19486b5a.png)

But it has two downsides:
1. Difficult to represent any additional properties
2. Doesn't work well for two setpoints (which I still find a bit confusing)

Feedback?

**Comment by mrstegeman** [comment]
I like the normal layout. I think the custom UI is going to get super confusing with the amount of possible properties (and differing enums) and such.

**Comment by mrstegeman** [comment]
@benfrancis Icons please?

**Comment by igilham** [comment]
Looks like a good start. It isn't clear from this issue how many different device types will be supported by this UI or if this is just an example to build from so it may be worth calling out some of the assumptions here.

It looks like you're assuming any thermostat will have the same features with similar names (auto, heating, cooling, off). My Nest thermostat only has to worry about **heating** and **eco** in normal use.

Some systems are only for heating or only for cooling. Some systems also have multiple set points rather than a target temperature. Some systems include ventilation options (fan speed etc.) and perhaps even heat recovery.

**Comment by mrstegeman** [comment]
The [Thermostat schema](https://iot.mozilla.org/schemas/#Thermostat) contained the most common set of functionality we could think of for a thermostat, and only 2 property types are actually required: Temperature and TargetTemperature.

We had a fair amount of discussion around different thermostats and what the support in the [GitHub issue](https://iot.mozilla.org/schemas/#Thermostat).

This UI and schema should allow most thermostats to be supported. Devices can always add additional properties to handle things like eco/vacation mode, filter warnings, etc.

**Comment by igilham** [comment]
> The [Thermostat schema](https://iot.mozilla.org/schemas/#Thermostat) contained the most common set of functionality we could think of for a thermostat, and only 2 property types are actually required: Temperature and TargetTemperature.
> 
> We had a fair amount of discussion around different thermostats and what the support in the [GitHub issue](https://github.com/mozilla-iot/schemas/issues/13).
> 
> This UI and schema should allow most thermostats to be supported. Devices can always add additional properties to handle things like eco/vacation mode, filter warnings, etc.

Cool, I appreciate the thoroughness.

**Comment by flatsiedatsie** [comment]
I've been implementing the thermostat capability in the MySensors add-on.

Here I have the `thermostat` property selected as the highlighted item. The thing also has a temperature sensor, making it the minimal viable thermostat capability.

![thermostat-or-temperature](https://user-images.githubusercontent.com/805405/68552553-241ba200-0410-11ea-99de-70943aa21b6a.png)

It looks as if it's showing the temperature property instead? Or am I missing something? I was hoping for a central bubble with the target temperature and (+) and (-) buttons to quickly be able to change it from the things overview page.

I'm also not seeing any up and down arrows in the thermostat bubble on the left? Normal number properties do at least have that?

**Comment by davehylands** [comment]
I would expect a Thermostat to show the current temperature, and that's what all of the real thermostats I've used do.

**Comment by flatsiedatsie** [comment]
But then why offer users the choice between showing the 'temperature' property and 'thermostat' property in the middle bubble if it doesn't make any difference? If I select the 'thermostat' for the center bubble, I expect to see the thermostat target temperature. If I wanted to see the actual temperature, I would have selected it.

All real thermostats also provide buttons to change the target temperature. Something like:
![thermostat_with_buttons_proposal](https://user-images.githubusercontent.com/805405/68589078-ec9a0d80-0482-11ea-96be-6b91a977e049.png)


Ideally, it would show both? After all, why have the capabilities, which have multiple minimally required properties, if this doesn't result in more powerful display options?
![thermostat_double_whammy](https://user-images.githubusercontent.com/805405/68589087-f15ec180-0482-11ea-9b0e-3e394cd30c12.png)

**Comment by flatsiedatsie** [comment]
Also, the thermostat icon alludes to there being up and down buttons to quickly change the temperature?
![thermostat-up-down-buttons](https://user-images.githubusercontent.com/805405/68591791-69c88100-0489-11ea-9527-7cec1b37dc6b.png)

But even in the big octopus view I can't find those buttons anywhere?
![thermostat-or-temperature](https://user-images.githubusercontent.com/805405/68591651-1e15d780-0489-11ea-8091-daf7d5ac1f2d.png)

It could be that I implemented something wrong. But for now it seems I've even lost the mini up-down arrows on the input field. It's like a string field input now, which makes it even less tablet/phone friendly :-( I hope I've just implemented it wrong, as usual ;-)

**Comment by mrstegeman** [comment]
@flatsiedatsie No production thermostat I've ever seen just shows the target temperature. They always show the current temperature as the primary display. In your second example, with both the target and current temperature, it's unclear to me which number is which. The primary display difference between this and a `TemperatureSensor` is that we show the `HeatingCoolingProperty` as well.

In your property breakout, the problem is most likely that you don't have a `minimum` and `maximum` for your `thermostat` property. In addition to those, you also need to set that property as `integer` or set a `multipleOf`. Without all of those things, the up/down arrows increment by insignificant amounts, so we omit them.

**Comment by flatsiedatsie** [comment]
Hmm, my code does actually provide the values you mention:
```
                if new_sub_type == 45: # V_HVAC_SETPOINT_HEAT
                    if "Temperature" in self._type:
                        self._type.remove('TemperatureSensor')
                    self._type.append('Thermostat')
                    self.properties[targetPropertyID] = MySensorsProperty(
                        self,
                        targetPropertyID,
                        {
                            '@type': 'TargetTemperatureProperty',
                            'label': new_description,
                            'minimum':0,
                            'maximum':150,
                            'type': 'number',
                            'unit': 'degree celsius',
                            'multipleOf':0.1,
                        },
                    values, new_value, new_node_id, new_child_id, new_main_type, new_sub_type)
```
Perhaps a 0.1 multipleOf is considered too small? But if you try to hatch chicken eggs or keep medicine at just the right temperature, you will care about that small an increment.

Back to the question: I agree no real physical thermostat does that. But this isn't a physical thermostat? It's a system that allows users to select what they want in the middle bubble. In my case, a user could even put something else entirely in the center bubble, not related to either the temperature or setpoint.

So my point is: if you want to see the temperature in the middle, the users can select that option. The temperature property is mandatory, after all.
But if users want to see the setpoint item in the main bubble, they should be able to do that? If in that case they want to see the temperature, they can open the octopus? Or ask via voice control?

Basically, this is weird:

Select temperature as central item -> see temperature
Select setpoint as central item -> see temperature

**Comment by mrstegeman** [comment]
I've fixed the up/down arrow issue: https://github.com/mozilla-iot/gateway/pull/2274

As for the rest:
* This is just a software representation of a physical device. If the physical device shows the current temperature by default, why should we not do the same?
* The user is selecting to show "Thermostat", not "setpoint".
* I'm hoping that in the near future, developers will be able to add/replace the web components used for the different capabilities with extensions.

**Comment by flatsiedatsie** [comment]
Ah, great! I was just going to post that even with conservative settings I couldn't get the arrows.

```
                if new_sub_type == 45: # V_HVAC_SETPOINT_HEAT
                    if "Temperature" in self._type:
                        self._type.remove('TemperatureSensor')
                    self._type.append('Thermostat')
                    self.properties[targetPropertyID] = MySensorsProperty(
                        self,
                        targetPropertyID,
                        {
                            '@type': 'TargetTemperatureProperty',
                            'label': new_description,
                            'minimum':0,
                            'maximum':30,
                            'type': 'number',
                            'unit': 'degree celsius',
                            'multipleOf':1,
                        },
```

**Comment by flatsiedatsie** [comment]
I suspect that the actual case, which I'm not seeing, is that the temperature is shown with the up and down buttons surrounding it? In that case, my point is less relevant. I don't care about seeing the current setpoint, I care about having the buttons.

> This is just a software representation of a physical device. If the physical device shows the current temperature by default, why should we not do the same?

I could also argue that all thermostats offer the ability to see the setpoint on them. But if I have to choose, then I also prefer seeing the temperature.

> The user is selecting to show "Thermostat", not "setpoint".

Interesting viewpoint, but it might be confusing? In all the other places the center bubble property (largely) corresponds to a property that can be seen in the octopus view? Which is then given center stage?

Put differently, this is my thought process as a user:
- I could see the setpoint in a bubble when I see the entire octopus.
- From experience, it seems I can select which of these bubbles I can give center stage.
- I assume the 'thermostat' option that I can select corresponds to the setpoint bubble (but you're right, it doesn't technically say setpoint).
- After believing I moved the setpoint bubble to the center position, it didn't show up in the same way way as in the octopus overview. -> confusion.

I'm a pretty long time user, and yet I can still be confused.

If you say "the center item doesn't have to correspond to a property bubble from the overview", then I say "ok, but in a lot of cases it does work like that, this would make the UI quite difficult to predict/quirky.

If you say "well, that's how it is", then I could ask "why can't I see the setpoint in a central bubble? If there is are no strict rules, couldn't it be a 'third' option?"

Perhaps to increase predictability of the UI, this could be the rule:
- A user can always put a bubble from the overview into the center position (I am still hoping that in a future update any property can take center stage, not just the ones from a capability).
- On top of that, we offer 'bonus' center bubbles that only appear when certain combinations of properties are available in a device. E.g. if there is a setpoint and a temperature bubble, then a new "thermostat" bubble becomes the bonus option.

**Comment by flatsiedatsie** [comment]
If I click on the up and down buttons of the center thermostat bubble, does it briefly show the new setpoint value in the place of the temperature? That would be a nice way of giving some feedback as to the new actual setpoint value.

**Comment by mrstegeman** [comment]
No, it doesn't. And yes, that's correct. It's just the up/down buttons for the input.

**Comment by flatsiedatsie** [comment]
Are there plans to implement bigger (mobile/tablet friendly) buttons to change the setpoint at some stage?

**Comment by mrstegeman** [comment]
No, not currently.

**Comment by flatsiedatsie** [comment]
Then I will brazenly create a feature request.
